cputest/fctiw.cpp

```diff
@@ -0,0 +1,208 @@
+  TowardsNegativeInfinity = 0b11
+};
+
+enum : u64
+{
+  DOUBLE_SIGN = 0x8000000000000000ULL,
+  DOUBLE_EXP = 0x7FF0000000000000ULL,
+  DOUBLE_FRAC = 0x000FFFFFFFFFFFFFULL,
+  DOUBLE_ZERO = 0x0000000000000000ULL,
+  DOUBLE_QBIT = 0x0008000000000000ULL
+};
+
+enum : u32
+{
+  FLOAT_SIGN = 0x80000000,
+  FLOAT_EXP = 0x7F800000,
+  FLOAT_FRAC = 0x007FFFFF,
+  FLOAT_ZERO = 0x00000000
+};
+
+// Algorithm adapted from Appendix C.4.2 in PowerPC Microprocessor Family:
+// The Programming Environments Manual for 32 and 64-bit Microprocessors
+static u64 fctiw_expected(double b, RoundingMode rounding_mode)
+{
+  const u64 upper_bits = 0xfff8000000000000ull;
+  if (std::isnan(b))
+    return upper_bits | 0x80000000;
+
+  const u64 bi = Common::BitCast<u64>(b);
+
+  s32 sign = static_cast<s32>((bi & DOUBLE_SIGN) >> 63);
+  s32 exp = static_cast<s32>((bi & DOUBLE_EXP) >> 52);
+  u64 frac = ((bi & DOUBLE_FRAC) << 11);
+
+  if (exp > 0)
+    frac |= 1ull << 63;
+
+  if (exp > 0)
+    exp -= 1023;
+  else
+    exp = -1022;
+
+  bool gbit = false;
+  bool rbit = false;
+  bool xbit = false;
+  for (s64 i = 0; i < 63 - exp; ++i)
+  {
+    xbit |= rbit;
+    rbit = gbit;
+    gbit = frac & 1;
+    frac >>= 1;
+  }
+
+  u32 inc = 0;
+  switch (rounding_mode)
+  {
+  case RoundingMode::Nearest:
+    if (gbit && ((frac & 1) || rbit || xbit))
+      inc = 1;
+    break;
+  case RoundingMode::TowardsZero:
+    // Nothing
+    break;
+  case RoundingMode::TowardsPositiveInfinity:
+    if (!sign && (gbit || rbit || xbit))
+      inc = 1;
+    break;
+  case RoundingMode::TowardsNegativeInfinity:
+    if (sign && (gbit || rbit || xbit))
+      inc = 1;
+    break;
+  }
+
+  frac += inc;
+
+  if (!sign && frac > 0x7fffffff)
+  {
+    // Positive large operand or +inf
+    frac = 0x7fffffff;
+  }
+  else if (frac > 0x80000000)
+  {
+    // Negative large operand or -inf
+    frac = 0x80000000;
+  }
+  else if (sign)
+  {
+    // Appendix C.4.2 does not cast to 32-bit here, but doing so matches
+    // Broadway's behavior of setting bit 31 to 1 for negative zeroes.
+    // Bits 0-31 are undefined according to appendix C.4.2.
+    frac = static_cast<u64>(~static_cast<u32>(frac)) + 1;
+  }
+
+  return upper_bits | frac;
+}
+
+static void FctiwTestIndividual(u32 i, RoundingMode rounding_mode)
+{
+  float input = Common::BitCast<float>(i);
+  u64 expected = fctiw_expected(input, rounding_mode);
+  u64 result = 0;
+  asm("fctiw %0, %1" : "=f"(result) : "f"(input));
+
+  DO_TEST(result == expected, "fctiwz 0x%08x (%f):\n"
+                              "     got 0x%016llx (%d)\n"
+                              "expected 0x%016llx (%d)",
+          i, input, result, static_cast<s32>(result), expected, static_cast<s32>(expected));
+}
+
+static void FctiwTestBothSigns(u32 i, RoundingMode rounding_mode)
+{
+  FctiwTestIndividual(i, rounding_mode);
+  FctiwTestIndividual(FLOAT_SIGN | i, rounding_mode);
+}
+
+// Float Convert To Integer Word
+static void FctiwTest()
+{
+  START_TEST();
```

Comment:
Could we use ScopeGuard here (or even add a ScopedTest wrapper for START_TEST / END_TEST) to avoid the gotos? :D

Reply:
Putting a copy of ScopeGuard in this repo would require us to bump the C++ version to C++17, which in turn causes compilation errors because we use the `register` keyword in one test... So that doesn't sound very fun.

ScopedTest sounds like a decent idea, but I think I'll do it in a different PR.

Reply:
It's fine to remove register in those cases. It was deprecated and removed because it does nothing in practice. Bumping the compiler requirements is fine.